File: internal/server/web/views/custom/panels/mtf_changer_grid.js
Ext.define("PBS.MtfManagement.ChangerGrid", {
  extend: "Ext.grid.Panel",
  alias: "widget.pbsMtfChangerGrid",

  controller: {
    xclass: "Ext.app.ViewController",

    onAdd: function () {
      let me = this;
      Ext.create("PBS.TapeManagement.ChangerEditWindow", {
        listeners: {
          destroy: function () {
            me.reload();
          },
        },
      }).show();
    },

    onEdit: function () {
      let me = this;
      let view = me.getView();
      let selection = view.getSelection();
      if (!selection || selection.length < 1) {
        return;
      }
      Ext.create("PBS.TapeManagement.ChangerEditWindow", {
        changerid: selection[0].data.name,
        autoLoad: true,
        listeners: {
          destroy: () => me.reload(),
        },
      }).show();
    },

    showStatus: function () {
      let view = this.getView();
      let selection = view.getSelection();
      if (!selection || selection.length < 1) {
        return;
      }
      location.hash = `#Changer-${encodeURIComponent(selection[0].data.name)}`;
    },

    onDblClick: function () {
      this.showStatus();
    },

    reload: function () {
      this.getView().getStore().rstore.load();
    },

    stopStore: function () {
      this.getView().getStore().rstore.stopUpdate();
    },

    startStore: function () {
      this.getView().getStore().rstore.startUpdate();
    },

    init: function (view) {
      Proxmox.Utils.monStoreErrors(view, view.getStore().rstore);
    },
  },

  listeners: {
    beforedestroy: "stopStore",
    deactivate: "stopStore",
    activate: "startStore",
    itemdblclick: "onDblClick",
  },

  store: {
    type: "diff",
    rstore: {
      type: "update",
      storeid: "proxmox-tape-changers",
      model: "pbs-model-changers",
      proxy: {
        type: "proxmox",
        url: "/api2/json/tape/changer",
        queryParam: null,
      },
    },
    sorters: "name",
  },

  tbar: [
    {
      text: gettext("Add"),
      xtype: "proxmoxButton",
      handler: "onAdd",
      selModel: false,
    },
    "-",
    {
      text: gettext("Edit"),
      xtype: "proxmoxButton",
      handler: "onEdit",
      disabled: true,
    },
    {
      text: gettext("Status"),
      xtype: "proxmoxButton",
      handler: "showStatus",
      disabled: true,
      iconCls: "fa fa-window-restore",
    },
    {
      xtype: "proxmoxStdRemoveButton",
      baseurl: "/api2/extjs/config/changer",
      callback: "reload",
    },
  ],

  columns: [
    {
      text: gettext("Name"),
      dataIndex: "name",
      flex: 1,
    },
    {
      text: gettext("Path"),
      dataIndex: "path",
      flex: 2,
    },
    {
      text: gettext("Vendor"),
      dataIndex: "vendor",
      flex: 1,
    },
    {
      text: gettext("Model"),
      dataIndex: "model",
      flex: 1,
    },
    {
      text: gettext("Serial"),
      dataIndex: "serial",
      flex: 1,
    },
    {
      text: gettext("Import/Export Slots"),
      dataIndex: "export-slots",
      flex: 1,
    },
  ],
});

File: internal/server/web/ui/mtf_changer_grid.go
package ui

import "github.com/pbs-plus/pbs-plus/internal/server/web/js"

var mtfChangerGrid = js.Grid{
	Name: "PBS.MtfManagement.ChangerGrid", XType: "pbsMtfChangerGrid",
	Store:     js.Store{StoreID: "proxmox-tape-changers", Model: "pbs-model-changers", APIPath: "/api2/json/tape/changer", Sorters: "name", Proxy: js.ProxyProxmox, QueryParamNull: true},
	Listeners: js.Listeners{ItemDblClick: "onDblClick"},
	Controller: js.Controller{Methods: map[string]js.Raw{
		"onAdd":      js.Func("", `let me = this; Ext.create("PBS.TapeManagement.ChangerEditWindow", { listeners: { destroy: () => me.reload() } }).show();`),
		"onEdit":     js.Func("", `let me = this; let selection = me.getView().getSelection(); if (!selection || selection.length < 1) return; Ext.create("PBS.TapeManagement.ChangerEditWindow", { changerid: selection[0].data.name, autoLoad: true, listeners: { destroy: () => me.reload() } }).show();`),
		"showStatus": js.Func("", `let selection = this.getView().getSelection(); if (!selection || selection.length < 1) return; location.hash = "#Changer-" + encodeURIComponent(selection[0].data.name);`),
		"onDblClick": js.Func("", `this.showStatus();`),
	}},
	Tbar: []js.Tool{
		{Text: "Add", Handler: "onAdd", SelModel: new(false)}, js.Sep(),
		{Text: "Edit", Handler: "onEdit", Disabled: true},
		{Text: "Status", Handler: "showStatus", Disabled: true, IconCls: "fa fa-window-restore"},
		{StandardRemoveBaseURL: "/api2/extjs/config/changer", Callback: "reload"},
	},
	Columns: []js.Column{
		{Text: "Name", DataIndex: "name", Flex: 1}, {Text: "Path", DataIndex: "path", Flex: 2},
		{Text: "Vendor", DataIndex: "vendor", Flex: 1}, {Text: "Model", DataIndex: "model", Flex: 1},
		{Text: "Serial", DataIndex: "serial", Flex: 1}, {Text: "Import/Export Slots", DataIndex: "export-slots", Flex: 1},
	},
}
